lib/src/source/git.dart

```diff
@@ -253,9 +302,22 @@ class GitSource extends CachedSource {
   /// by [id] on the effective ref of [id].
   ///
   /// This assumes that the canonical clone already exists.
-  Future<String> _getRev(PackageId id) {
-    return git.run(["rev-list", "--max-count=1", _getEffectiveRef(id)],
-        workingDir: _repoCachePath(id)).then((result) => result.first);
+  Future<String> _getRev(PackageId id) async {
+    var ref = _getEffectiveRef(id);
```

Comment:
`id.version` ends up being set to the version number in the pubspec.yaml after the solver has selected a pubspec: https://github.com/dart-lang/pub/blob/master/lib/src/solver/version_solver.dart#L257

`resolveId` is called once again after a solution has been reached: https://github.com/dart-lang/pub/blob/master/lib/src/solver/backtracking_solver.dart#L168. `resolveId` uses `_ensureRevision` which uses `_getRev`.

At this point `id.version` has something even if one wasn't specified in the consumer's pubspec.

So this is why `if (ref == null && id.version != null && id.version != Version.none)` is not a sufficient condition to determine whether or not we should do a `rev-list $version`.

Open to suggestions :)